research/2d-decoding/rectangle_detector.py

```diff
@@ -0,0 +1,220 @@
+                            break
+                    else: # If no pixels within the tolerance are white, then the line isn't continuous, so this group isn't a rectangle
+                        nottaSquare = True
+                        break
+            else:
+                for y in range(minX[1], maxY[1]+1):
+                    for offset in range(-LINE_PIXEL_TOLERANCE, LINE_PIXEL_TOLERANCE+1):
+                        if minX[0]+offset < 0 or minX[0]+offset >= len(pixels[y]):
+                            continue
+                        if pixels[y][minX[0]+offset] == 0:
+                            break
+                    else:
+                        nottaSquare = True
+                        break
+        except ZeroDivisionError:
+            continue
+        if nottaSquare:
+            continue
+
+        try:
+            line = lambda a: (minX[1]-minY[1])/(minX[0]-minY[0])*(a - minX[0]) + minX[1] # Upper left side
+            for x in range(minX[0], minY[0]+1):
+                y = round(line(x))
+                for offset in range(-LINE_PIXEL_TOLERANCE, LINE_PIXEL_TOLERANCE+1):
+                    if y+offset < 0 or y+offset >= len(pixels):
+                        continue
+                    if pixels[y+offset][x] == 0:
+                        break
+                else: # If no pixels within the tolerance are white, then the line isn't continuous, so this group isn't a rectangle
+                    nottaSquare = True
+                    break
+        except ZeroDivisionError:
+            continue
+        if nottaSquare:
+            continue
+        
+        try:
+            if minY[0]-maxX[0] != 0:
+                line = lambda a: (minY[1]-maxX[1])/(minY[0]-maxX[0])*(a - maxX[0]) + maxX[1] # Upper right side
+                for x in range(minY[0], maxX[0]+1):
+                    y = round(line(x))
+                    for offset in range(-LINE_PIXEL_TOLERANCE, LINE_PIXEL_TOLERANCE+1):
+                        if y+offset < 0 or y+offset >= len(pixels):
+                            continue
+                        if pixels[y+offset][x] == 0:
+                            break
+                    else: # If no pixels within the tolerance are white, then the line isn't continuous, so this group isn't a rectangle
+                        nottaSquare = True
+                        break
+            else:
+                for y in range(minY[1], maxX[1]+1):
+                    for offset in range(-LINE_PIXEL_TOLERANCE, LINE_PIXEL_TOLERANCE+1):
+                        if minY[0]+offset < 0 or minY[0]+offset >= len(pixels[y]):
+                            continue
+                        if pixels[y][minY[0]+offset] == 0:
+                            break
+                    else:
+                        nottaSquare = True
+                        break
+        except ZeroDivisionError:
+            continue
+        if nottaSquare:
+            continue
```

Comment:
There must be a better way to do this, this is basically copy paste with some variable name changes

Reply:
@TTG3333 Any ideas on how to simplify it? I propose as part of your PR documenting the algorithm, you can comment this section more as well to make it more obvious to external observers what's going on (instead of `# Lower right side`, you could explain what's being calculated here and how it's being used later in the code).

Reply:
Fixed by TTG in a5bb373